### PR TITLE
fix: Add migration to fix NotificationServicesController bug

### LINKED ISSUE #12207: TypeError: undefined is not an object (evaluating 't.metamaskNotificationsList')
Sentry Issue: [METAMASK-MOBILE-31YF](https://metamask.sentry.io/issues/6044179124/?referrer=github_integration)

```
TypeError: undefined is not an object (evaluating 't.metamaskNotificationsList')
  at None (app:///main.jsbundle:4914:2512)
  at E (app:///main.jsbundle:2878:564)
  at None (app:///main.jsbundle:2877:1041)
  at E (app:///main.jsbundle:2878:564)
  at None (app:///main.jsbundle:4913:347)
...
(11 additional frame(s) were not displayed)
```

### PR DESCRIPTION
**Pull request by smilingkylan**
## **Description**

The purpose of this PR is to fix an error apparently caused by a missing `NotificationServicesController` property on the `state.engine.backgroundState`

## **Related issues**

Fixes: #12207 

## **Manual testing steps**

Although we didn't pinpoint what was causing the `NotificationServicesController` to be missing, this code fills in the data if it is missing.

## **Screenshots/Recordings**

n/a

### **Before**

n/a

### **After**

n/a

## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

File: app/store/migrations/060.ts [hunk 1]
```diff
@@ -0,0 +1,16 @@
+export default function migrate(state) {
+    state.engine.backgroundState.NotificationServicesControllerState = undefined
```
Comment (tommasini) [inline app/store/migrations/060.ts:2]:
```suggestion
  if( !state.engine.backgroundState.NotificationServicesControllerState )
```

File: app/store/migrations/060.ts [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+        isMetamaskNotificationsFeatureSeen: false,
+        isNotificationServicesEnabled: false,
+        isUpdatingMetamaskNotifications: false,
+        isUpdatingMetamaskNotificationsAccount: [],
+        metamaskNotificationsList: [],
+        metamaskNotificationsReadList: [],
+        subscriptionAccountsSeen: [],
+      }
+    return state;
+}
```
Comment (tommasini) [inline app/store/migrations/060.ts:16]: As a unit test case we can add if the NotificationServicesController is undefined before applying the migration, it should be defined after

we should also cover the `ensureValidState` function test cases with this 3 scenarios
```
  {
      state: null,
      errorMessage: "FATAL ERROR: Migration 58: Invalid state error: 'object'",
      scenario: 'state is invalid',
    },
    {
      state: merge({}, initialRootState, {
        engine: null,
      }),
      errorMessage:
        "FATAL ERROR: Migration 58: Invalid engine state error: 'object'",
      scenario: 'engine state is invalid',
    },
    {
      state: merge({}, initialRootState, {
        engine: {
          backgroundState: null,
        },
      }),
      errorMessage:
        "FATAL ERROR: Migration 58: Invalid engine backgroundState error: 'object'",
      scenario: 'backgroundState is invalid',
    },
```

File: app/store/migrations/060.ts [hunk 3]
```diff
@@ -0,0 +1,16 @@
+export default function migrate(state) {
```
Comment (tommasini) [inline app/store/migrations/060.ts:1]:
```suggestion
export default function migrate(state: unknown) {
```

File: app/store/migrations/060.ts [hunk 4]
```diff
@@ -0,0 +1,20 @@
+import { ensureValidState } from "./util";
+
+export default function migrate(state: unknown) {
+    if (!ensureValidState(state, 60)) {
+        return state;
+      }
+    state.engine.backgroundState.NotificationServicesController = {
```
Comment (Cal-L) [inline app/store/migrations/060.ts:7]: We only want to set `NotificationServicesController` if it is undefined. Take a look at migration 59 for reference. Similarly, you can reference those unit tests as well to write unit tests for this migration.

**Comment by tommasini** [comment]
Let's also add the verification to the getTags file
add this checks [here](https://github.com/MetaMask/metamask-mobile/blob/475e9bfe12e25ef9183a1170a5d6510e2cad608a/app/util/sentry/tags/index.ts#L11) if you have the time please
```
 if (!state?.engine?.backgroundState?.AccountsController) return;
  if (!state?.engine?.backgroundState?.NftController) return;
  if (!state?.engine?.backgroundState?.NotificationServicesController) return;
  if (!state?.engine?.backgroundState?.TokensController) return;
  if (!state?.engine?.backgroundState?.TransactionController) return;
```

File: app/store/migrations/060.ts (hunk 4 above)
Comment (kylanhurt) [reply app/store/migrations/060.ts]: @Cal-L Let me know if the new conditional is better

File: app/store/migrations/060.test.ts [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+      );
+    });
+  }
+
+  it('should insert default NotificationServicesController if missing', async () => {
+    const oldState = {
+      engine: {
+        backgroundState: {},
+      },
+    };
```
Comment (tommasini) [inline app/store/migrations/060.test.ts:65]:
```suggestion
    const oldState = {
    ...initialRootState,
      engine: {
        backgroundState: {},
      },
    };
```

File: app/store/migrations/060.test.ts [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+
+    const expectedState = {
+      engine: {
+        backgroundState: {
+            NotificationServicesController: DEFAULT_NOTIFICATION_SERVICES_CONTROLLER
+        },
+      },
+    };
```
Comment (tommasini) [inline app/store/migrations/060.test.ts:73]:
```suggestion
    const expectedState = {
    ...initialRootState,
      engine: {
        backgroundState: {
            NotificationServicesController: DEFAULT_NOTIFICATION_SERVICES_CONTROLLER
        },
      },
    };
```

**Review by tommasini — APPROVED** [review]

**Review by sethkfman — APPROVED** [review]
LGTM

**Pull request merged**